crypto/evp/evp_lib.c

```diff
@@ -11,20 +11,37 @@
 #include "internal/cryptlib.h"
 #include <openssl/evp.h>
 #include <openssl/objects.h>
+#include <openssl/params.h>
+#include <openssl/core_names.h>
 #include "internal/evp_int.h"
 #include "internal/provider.h"
 #include "evp_locl.h"
 
 int EVP_CIPHER_param_to_asn1(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
 {
     int ret;
+    const EVP_CIPHER *cipher = c->cipher;
+
+    if (cipher->prov != NULL) {
+        /*
+         * The cipher has come from a provider and won't have the default flags.
+         * Find the implicit form so we can check the flags.
+         * TODO(3.0): This won't work for 3rd party ciphers we know nothing about
```

Comment:
We'll still be able to look it up by NID won't we?
I think supporting NID-less algorithms isn't sensible, have something new create an object.

Reply:
> We'll still be able to look it up by NID won't we?
I think supporting NID-less algorithms isn't sensible, have something new create an object.

NIDs are added automatically for algorithms we've not encountered before. We should be able to support such algorithms (consider GOST). We've not yet done anything to EVP_get_cipherbynid() for ciphers with NIDs added at runtime but in any case such ciphers won't have the right flags needed for this function.

Reply:
Okay.  Several places do lookups by NID in some subtly different ways -- I'd like unify and simplify them.  EVP_get_cipherbynid() would be one such.